python/ray/serve/_private/http_state.py

```diff
@@ -72,65 +76,136 @@ def set_status(self, status: HTTPProxyStatus) -> None:
         self._status = status
         self.update_actor_details(status=self._status)
 
+    def try_update_status(self, status: HTTPProxyStatus):
+        """Try update with the new status and only update when the conditions are met.
+
+        Status will only set to UNHEALTHY after PROXY_HEALTH_CHECK_UNHEALTHY_THRESHOLD
+        consecutive failures. A warning will be logged when the status is set to
+        UNHEALTHY. Also, when status is set to HEALTHY, we will reset
+        self._consecutive_health_check_failures to 0.
+        """
+        # Early return to skip setting UNHEALTHY status if there are still room for
+        # retry.
+        if (
+            status == HTTPProxyStatus.UNHEALTHY
+            and self._consecutive_health_check_failures
+            < PROXY_HEALTH_CHECK_UNHEALTHY_THRESHOLD
+        ):
+            self._consecutive_health_check_failures += 1
+            logger.info(
+                f"HTTP proxy {self._actor_name} failed the health check "
+                f"{self._consecutive_health_check_failures} times in a row, not marking"
+                " it unhealthy just yet."
+            )
+            return
+
+        # Reset self._consecutive_health_check_failures when status is set to HEALTHY.
+        if status == HTTPProxyStatus.HEALTHY:
+            self._consecutive_health_check_failures = 0
+
+        self.set_status(status=status)
+
+        # If all retries have been exhausted and setting the status to UNHEALTHY, log a
+        # warning message to the user.
+        if (
+            self._consecutive_health_check_failures
+            >= PROXY_HEALTH_CHECK_UNHEALTHY_THRESHOLD
+        ):
```

Comment:
I think this check is redundant with the early return above. You can just do `if status == UNHEALTHY` here.

Would also suggest to only log on state transition to `UNHEALTHY` else this will spam multiple times.

Reply:
The way this is written is probably a bit confusing. But basically the one above with early return is attempting to set unhealthy, but did not happen due to there are still retries left. We just log a warning in this case. This one here will only be triggered when the status is already set to unhealthy (if it's set to healthy, `self. _consecutive_health_check_failures` will be reset to 0 and will not go into this case). But I do see that it can be spammy for the retry logging and can make it more readable by just check the status here. Great suggestions! Will push a fix